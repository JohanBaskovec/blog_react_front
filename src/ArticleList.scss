.ArticleList {
  display: flex;
  flex-direction: column;
}
