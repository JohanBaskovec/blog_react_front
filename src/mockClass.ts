

type constructorTypeOf<T> = new (...args:any[]) => T;

export function mockDefaultApi<T>(o: T): T {
    const mocked: any = {};
    for (const key in o) {
        if (typeof o[key] === 'function') {
            mocked[key] = jest.fn();
        }
    }
    return mocked;
}

type MockedProxy<T> = jest.Mock<T extends (...args: any[]) => infer U ? U : T, any[]>;

type MockedClass<T> = {
    [k in keyof T]: MockedProxy<T[k]>;
}

export function mockClass<T>(constr: constructorTypeOf<T>): T & MockedClass<T>{
    let obj: T = new constr();
    const mocked: any = {};

    do {
        const properties = Object.getOwnPropertyNames(obj);
        for (const propertyName of properties) {
            if (typeof (obj as any)[propertyName] === "function" && !mocked[propertyName]) {
                mocked[propertyName] = jest.fn();
            }
        }
    } while (obj = Object.getPrototypeOf(obj));
    return mocked as any;
}
