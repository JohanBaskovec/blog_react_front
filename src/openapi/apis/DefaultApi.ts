// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { Observable } from 'rxjs';
import { BaseAPI, HttpHeaders, throwIfNullOrUndefined, encodeURI } from '../runtime';
import {
    Article,
    ArticleFormData,
    LoginForm,
    RegistrationForm,
    User,
} from '../models';

export interface GetArticleByIdRequest {
    id: string;
}

export interface InsertArticleRequest {
    articleFormData: ArticleFormData;
}

export interface LoginRequest {
    loginForm: LoginForm;
}

export interface RegisterRequest {
    registrationForm: RegistrationForm;
}

export interface UpdateArticleRequest {
    articleFormData: ArticleFormData;
}

/**
 * no description
 */
export class DefaultApi extends BaseAPI {

    /**
     * Get all the articles
     */
    getAllArticles = (): Observable<Array<Article>> => {
        return this.request<Array<Article>>({
            path: '/article',
            method: 'GET',
        });
    };

    /**
     * Get an article
     */
    getArticleById = ({ id }: GetArticleByIdRequest): Observable<Article> => {
        throwIfNullOrUndefined(id, 'getArticleById');

        return this.request<Article>({
            path: '/article/{id}'.replace('{id}', encodeURI(id)),
            method: 'GET',
        });
    };

    /**
     * Get the current session
     */
    getCurrentAuthenticatedUser = (): Observable<User> => {
        const headers: HttpHeaders = {
        };

        return this.request<User>({
            path: '/me',
            method: 'GET',
            headers,
        });
    };

    /**
     * Insert an article
     */
    insertArticle = ({ articleFormData }: InsertArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'insertArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'POST',
            headers,
            body: articleFormData,
        });
    };

    /**
     * Create a new HTTP session
     */
    login = ({ loginForm }: LoginRequest): Observable<User> => {
        throwIfNullOrUndefined(loginForm, 'login');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<User>({
            path: '/login',
            method: 'POST',
            headers,
            body: loginForm,
        });
    };

    /**
     * Logout
     */
    logout = (): Observable<void> => {
        const headers: HttpHeaders = {
        };

        return this.request<void>({
            path: '/logout',
            method: 'POST',
            headers,
        });
    };

    /**
     * Register
     */
    register = ({ registrationForm }: RegisterRequest): Observable<void> => {
        throwIfNullOrUndefined(registrationForm, 'register');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/register',
            method: 'POST',
            headers,
            body: registrationForm,
        });
    };

    /**
     * Update an article
     */
    updateArticle = ({ articleFormData }: UpdateArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'updateArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'PUT',
            headers,
            body: articleFormData,
        });
    };

}
