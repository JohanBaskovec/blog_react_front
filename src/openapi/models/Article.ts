// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface Article
 */
export interface Article {
    /**
     * @type {string}
     * @memberof Article
     */
    id: string;
    /**
     * @type {string}
     * @memberof Article
     */
    title: string;
    /**
     * @type {string}
     * @memberof Article
     */
    content: string;
}
