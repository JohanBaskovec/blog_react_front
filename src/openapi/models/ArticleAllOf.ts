// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import {
    User,
} from './';

/**
 * @export
 * @interface ArticleAllOf
 */
export interface ArticleAllOf {
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    title: string;
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    content: string;
    /**
     * @type {User}
     * @memberof ArticleAllOf
     */
    author: User;
    /**
     * @type {number}
     * @memberof ArticleAllOf
     */
    version: number;
}
