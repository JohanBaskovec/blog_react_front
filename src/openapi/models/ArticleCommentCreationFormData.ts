// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface ArticleCommentCreationFormData
 */
export interface ArticleCommentCreationFormData {
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    id: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    articleId: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    content: string;
}
