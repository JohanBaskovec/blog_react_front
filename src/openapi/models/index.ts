export * from './Article';
export * from './ArticleAllOf';
export * from './ArticleFormData';
export * from './Authorization';
export * from './ClientError';
export * from './HttpSession';
export * from './LoginForm';
export * from './ObjectBase';
export * from './Permission';
export * from './PermissionAuthorization';
export * from './PermissionAuthorizationAllOf';
export * from './RegistrationForm';
export * from './Role';
export * from './RoleAuthorization';
export * from './RoleAuthorizationAllOf';
export * from './ServerError';
export * from './User';
export * from './ValidationError';
