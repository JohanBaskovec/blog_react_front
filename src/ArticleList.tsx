import React, {useContext, useEffect, useState} from "react";
import {Article} from "./openapi/models";
import {DefaultApi} from "./openapi/apis";
import {ArticleSummary} from "./ArticleSummary";
import "./ArticleList.scss";
import {AppLink} from "./AppLink";
import {SessionContext} from "./SessionContext";

export interface ArticleListProps {
    api: DefaultApi;
}

export function ArticleList({api}: ArticleListProps) {
    const [articles, setArticles] = useState<Article[]>([]);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<any>(null);
    useEffect(() => {
        setLoading(true);
        const getAllArticles$ = api.getAllArticles()
            .subscribe(articles => {
                    setArticles(articles);
                    setLoading(false);
                },
                (err) => {
                    setLoading(false);
                    setError(err);
                });
        return () => {
            getAllArticles$.unsubscribe();
        }
    }, []);
    const articleSummaryMargin = "1rem 0";
    const articleSummaryStyle = {
        margin: articleSummaryMargin,
    };
    const session = useContext(SessionContext);
    let body = null;

    if (loading) {
        body = <>Loading articles...</>;
    } else if (error) {
        const errorResponse = error.response ? error.response.message : "Unknown error.";
        body = <>An error happened: {errorResponse}</>;
    } else {
        body = articles.map((article) => <ArticleSummary key={article.id}
                                                         article={article}
                                                         style={articleSummaryStyle}/>);
    }
    return (
        <div className="ArticleList">
            {session.user ?
                <div>
                    <AppLink to="/article/new">New article</AppLink>
                </div> :
                null
            }
            {body}
        </div>
    )
}
