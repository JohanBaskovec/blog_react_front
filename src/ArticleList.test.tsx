import React from "react";
import {DefaultApi} from "./openapi/apis";
import {of} from "rxjs";
import {mockClass} from "./mockClass";
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import pretty from "pretty";
import {ArticleList} from "./ArticleList";
import {Article} from "./openapi/models";

let container: HTMLDivElement;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    if (container != null) {
        unmountComponentAtNode(container);
        container.remove();
    }
});

describe('ArticleList', () => {
    it("renders the article list when there are no articles", () => {
        const api: DefaultApi = mockClass(new DefaultApi());
        api.getAllArticles = jest.fn(() => {
            const articles: Array<Article> = [];
            return of(articles);
        });
        act(() => {
            render(<ArticleList api={api}/>, container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });

    it("renders the article list when there is 1 article", () => {
        const api: DefaultApi = mockClass(new DefaultApi());
        api.getAllArticles = jest.fn(() => {
            const articles: Array<Article> = [
                {
                    id: "id",
                    title: "title",
                    content: "content",
                },
            ];
            return of(articles);
        });
        act(() => {
            render(<ArticleList api={api}/>, container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });

    it("renders the article list when there are 3 articles", () => {
        const api: DefaultApi = mockClass(new DefaultApi());
        api.getAllArticles = jest.fn(() => {
            const articles: Array<Article> = [
                {
                    id: "id0",
                    title: "title0",
                    content: "content0",
                },
                {
                    id: "id1",
                    title: "title1",
                    content: "content1",
                },
                {
                    id: "id2",
                    title: "title2",
                    content: "content2",
                },
            ];
            return of(articles);
        });
        act(() => {
            render(<ArticleList api={api}/>, container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
});
