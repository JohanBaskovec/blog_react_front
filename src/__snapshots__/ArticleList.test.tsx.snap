// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList renders the article list when there are 3 articles and user is logged in 1`] = `
"<div class=\\"ArticleList\\">
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id0\\">title0</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content0</div>
  </div>
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id1\\">title1</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content1</div>
  </div>
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id2\\">title2</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content2</div>
  </div>
</div>"
`;

exports[`ArticleList renders the article list when there are 3 articles and user is logged out 1`] = `
"<div class=\\"ArticleList\\">
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id0\\">title0</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content0</div>
  </div>
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id1\\">title1</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content1</div>
  </div>
  <div class=\\"ArticleSummary\\" style=\\"margin: 1rem 0px;\\">
    <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue\\" href=\\"/article/id2\\">title2</a></div>
    <div class=\\"ArticleSummary__publication_time\\">Published on Dec 19, 2020, 5:48 AM</div>
    <div class=\\"ArticleSummary__author\\">by username</div>
    <div class=\\"ArticleSummary__content\\">content2</div>
  </div>
</div>"
`;
